tests/output/test_hdf5file_checkpoint.py

```diff
@@ -0,0 +1,125 @@
+
+
+@pytest.fixture(params=range(1, 4))
+def degree(request):
+    return request.param
+
+
+@pytest.fixture(params=["CG"])
+def fs(request):
+    return request.param
+
+
+@pytest.fixture
+def dumpfile(tmpdir):
+    return str(tmpdir.join("dump.h5"))
+
+
+@pytest.fixture(scope="module")
+def f():
+    m = UnitSquareMesh(2, 2)
+    V = FunctionSpace(m, 'CG', 1)
+    f = Function(V, name="f")
+    f.interpolate(Expression("x[0]*x[1]"))
+    return f
+
+
+def run_store_load(mesh, fs, degree, dumpfile):
+
+    V = FunctionSpace(mesh, fs, degree)
+
+    f = Function(V, name="f")
+
+    expr = Expression("x[0]*x[1]")
+
+    f.interpolate(expr)
+
+    f2 = Function(V, name="f")
+
+    dumpfile = mesh.comm.bcast(dumpfile, root=0)
+
+    with HDF5File(dumpfile, "w", comm=mesh.comm) as h5:
+        h5.write(f, "/solution")
+        h5.read(f2, "/solution")
+
+    assert np.allclose(f.dat.data_ro, f2.dat.data_ro)
+
+    with HDF5File(dumpfile, "w", comm=mesh.comm) as h5:
+        h5.write(f, "/solution", timestamp=math.pi)
+        h5.read(f2, "/solution", timestamp=math.pi)
+
+    assert np.allclose(f.dat.data_ro, f2.dat.data_ro)
+
+def test_checkpoint_fails_for_non_function(dumpfile):
+    dumpfile = MPI.COMM_WORLD.bcast(dumpfile, root=0)
+    with HDF5File(dumpfile, "w", comm=MPI.COMM_WORLD) as h5:
+        with pytest.raises(ValueError):
+            h5.write(np.arange(10), "/solution")
+
+
+def test_store_load(mesh, fs, degree, dumpfile):
+    run_store_load(mesh, fs, degree, dumpfile)
+
+
+@pytest.mark.parallel(nprocs=2)
+def test_store_load_parallel(mesh, fs, degree, dumpfile):
+    run_store_load(mesh, fs, degree, dumpfile)
+
+
+def test_checkpoint_read_not_exist_ioerror(dumpfile):
+    with pytest.raises(IOError):
+        with HDF5File(dumpfile, file_mode="r"):
+            pass
+
+
+def test_attributes(f, dumpfile):
+    mesh = f.function_space().mesh()
+    dumpfile = mesh.comm.bcast(dumpfile, root=0)
+    with HDF5File(dumpfile, file_mode="w", comm=mesh.comm) as h5:
+        with pytest.raises(KeyError):
+            attrs = h5.attributes("/foo")
+            attrs["nprocs"] = 1
+        with pytest.raises(KeyError):
+            attrs = h5.attributes("/bar")
+            attrs["nprocs"]
+
+        h5.write(mesh.coordinates, "/coords")
+        attrs = h5.attributes("/coords")
+        attrs["dimension"] = mesh.coordinates.dat.cdim
+
+        assert attrs["dimension"] == mesh.coordinates.dat.cdim
+
+def test_store_read_only_ioerror(f, dumpfile):
+    # Make file
+    with HDF5File(dumpfile, "w") as h5:
+        pass
+    with HDF5File(dumpfile, "r") as h5:
+        with pytest.raises(IOError):
+            h5.write(f, "/function")
+
+def test_multiple_timestamps(f, dumpfile):
+    with HDF5File(dumpfile, "w") as h5:
+        h5.write(f, "/solution", timestamp=0.1)
+        h5.write(f, "/solution", timestamp=0.2)
+
+        timestamps = h5.get_timestamps()
+
+        assert np.allclose(timestamps, [0.1, 0.2])
```

Comment:
Can this test write two different values at the timesteps and assert that you can read them back in?